### DISCUSSION TITLE
`設定項目` 的格式與可擴展性

### DISCUSSION THREAD

**Discussion by syimyuzya** [nk2028/tshet-uinh-deriver-tools#4] [Ideas]
昨天 @graphemecluster 為推導器 v0.2 的參數名與分組標籤加入了 `description` 的支持，效果是鼠標放在選項上時可以彈出其詳細說明（在參數項目中的指定方式是 `["<key>\n<description>", <default value (& available options)>]`，分組標籤格式亦類似）。

這是一個極其有用，而且不論 GUI、CLI（若有的話）都適合加入的功能。因此 deriver-tools 自然也要加入其支持～ww【當然，考慮到原本的指定格式擴展性很受限，deriver-tools 不會用這種格式，而是會重新設計一下。

同時，受此啟發，我還想到之後 GUI 信息的部分是不是還可以繼續加強，比如把參數的「key」（JS 中用的屬性名，JS/GUI 通用）和「名稱」（顯示於推導器界面，GUI 專用）也分開，以及參數若提供可選項，每個項是不是也可以分別指定「value」（JS 用）與「text」（GUI 用）等等……

為了將來的可擴展性考慮，現在可能已經到了要好好設計推導方案中指定 `設定項目` 的格式的時候了。

## 設計目標

1. 以 JS API 為基本，GUI 擴展屬性作為輔助用（依託腳本批量化處理是 nk2028 的重大優勢（沒有之一😃），讓 GUI 一側變豐富卻令 JS API 一側要顧慮的東西增加，就得不償失了（當初做 `Qieyun.推導方案` 也正是為了解決這個（
2. 需要考慮未來可擴展性，符合「open-closed 原則」（即未來做到允許 GUI（或其他調用推導方案的程序）自行定義擴展屬性，同時不需修改已有的列表格式規範）
3. 讓推導方案仍能像現在一樣簡單地寫出列表來，不需要太多 boilerplate

## 方案側的 `參數項目` 格式設計思路

模仿一些動態語言（如 Python）中 positional args 與 keyword args 相結合的模式，讓用戶既能直接以 tuple 形式指定必要或常用的屬性，也能輔以一個 mapping 來指定個別屬性。而解析後的結果則會統一用 object 格式。

在 JS 裡，格式則類似於 `[arg1, arg2, ..., { kwarg1: ..., kwarg2: ... }]`，若不想指定某個 positional arg，可以寫 `undefined` [^kwargs]

具體來說，方案指定參數項目時：

- `key` 和 `default`：必要屬性，一般只用 positional 指定，即位於 tuple 的前兩項（雖然用 keyword 指定也是可以的）
- `options`：常用屬性，可以置於 positional 第三項（或 keyword 指定，下同）
  - `options` 中的各項可選值也可以詳細指定，見後述
- `name`：GUI 為用戶顯示（或 CLI 為用戶提示）的名稱（若和 key 不同的話）是常用屬性，置於第四項
- `description` 則置於第五項
- 更多自定義擴展屬性等，則僅可用 keyword 指定

【注意：把 `default` 和 `options` 拆開，且 index 也改為從 0 計，這是一個不兼容更動，原因是這樣可以允許程序讀取預設值時，不必關心可選值的存在與否；當然如果需要的話，也可以額外增設兼容過去寫法的 overload

而在解析之後，`設定項目` 均會被解析為 object 形式，以易於處理。

### 可選值的擴展屬性

一般情況下，可選值直接寫一個（不是 Array 或其他 object 的）值即可，在需要擴展屬性時，可以寫 tuple 或 object；最終均會被解析成 object：

- `value`：必要屬性，即該項的值
- `text`：置於第二項（或 keyword 指定），為 GUI 為用戶顯示（或 CLI 為用戶提示）的內容（若和 value 不同的話）
- 其他自定義擴展屬性均用 keyword 指定

### 「特殊項目」：並非參數項目的設定項目

若 tuple 的第一項以 `$` 開頭，則不視為參數項目（同時 keyword 下要用 `type` 而非 `key` 指定）。目前準備定義如下一些：

- `$label`：在解析後的列表中會作為一個分組標籤對象（方案中可簡寫為一個非空的串），GUI/CLI 會依此為用戶呈現分組後的參數（推導器目前表現為換行並寫上該分組的名稱）
- `$sep`：在解析後的列表中會作為一個分隔符對象（方案中可用 null 或空串指定），表明這是一個「分隔符」，推導器將其呈現為換行
- `$legacy`（準備棄用）：meta 信息，不會出現在解析後的列表中，表明方案基於 Qieyun.js v0.12 的音韻地位體系。從 Qieyun.js（TshetUinh.js？）v0.14 開始將不再完整支持（會出現部分行為差異），並將在 v0.15 徹底取消（也可能提前到 v0.14 就取消ww
- `$ext`（暫緩實現）：在解析後的列表中會作為一個擴展項目對象，內容為自定義對象

### 分組標籤的擴展屬性

分組標籤通常可以直接寫一個 string。在需要擴展屬性時，可寫作 tuple，首項固定為 `'$label'`。

- `text`：第二項，表示標籤名
- `description`：第三項，詳細說明
- 其他自定義擴展屬性需用 keyword 指定

## 不正確數據的處理方法與錯誤回報

- tuple 第一項不是 string
- 重複的參數 `key`
- 重複的可選值 `value`
- 項目缺少必要屬性
- `defaultValue` 不在 `options` 範圍內
- Positional 和 keyword 重複指定相同屬性
- 有相衝突屬性

無法處理的錯誤內容會被忽略，並將警告信息保存於 `推導設定.解析錯誤` 中。

### 為提供的選擇項指定不同的 value 與 text：

```js
// 沒有 value-text 分離時：
[
  ['版本', [2, '2020：切韻擬音 J（2022 新版）', '2022：切韻擬音 L'],
]
...
if (選項.版本 === '2020：切韻擬音 J（2022 新版）') ... // :(

// 有分離時：
[
  ['版本', 'L', [
    ['J2022', '2020：切韻擬音 J（2022 新版）'],
    ['L', '2022：切韻擬音 L'],
  ]],
]
...
if (選項.版本 === 'J2022') ... // :)
```

### 為參數加上 hidden 屬性

當推導器識別該屬性時，會不允許 GUI 用戶調整該參數，但 JS 用戶通過 API 調用時仍能享受「自動填入預設值」的功能。

```js
[
  ['模式', 0, ['嚴格', '標準', '寬鬆'], { hidden: true }],
  ['脣音咍韻推導為開口', false, { hidden: true }],
]
```

[^kwargs]: 判斷用戶有沒有給出 kwargs 一項的方式是看 tuple 最後一項是不是 object 且非 Array；但這樣若用戶提供的最後一項 positional arg 恰好是個 object，則會被誤當成 kwargs；這種情況很鮮見，即使遇到也不難解決：在其後再加一個空對象 `{}` 作為 kwargs 即可。

**Comment by graphemecluster** [comment]
<details>​
<summary>顯示構思中內容</summary>

我在想，以 TypeScript 型定義和 JSDoc 註釋為基礎的設定項目是否可行。
這樣 API 用戶既可以在編寫參數時獲得提示，又能受惠於 TypeScript 的型檢查；GUI 方面，monaco-editor 內建的 TypeScript worker 提供了豐富的型類別 API，我們只要利用 API 獲取這些型類別信息便能構建適當的表單欄位。

```ts
abstract class 參數項目 {
  $legacy?: boolean;
  // ...
}

class 推導方案<T> {
  constructor(private readonly 原始推導函數: (地位: 音韻地位, 字頭: string | null, 取得選項: <屬性 extends 參數項目>(設定: new () => 屬性) => 屬性) => T) {}
  推導(選項: InstanceType<Extract<T, new () => 參數項目>>): Exclude<T, new () => 參數項目> {}
}

new 推導方案((地位, 字頭, 取得選項) => {
  class 設定 extends 參數項目 {
    /**
     * 描述形式
     * （説明）
     */
    form: "描述" | "最簡描述" = "描述";
  }
  if (!地位) return 設定;
  const { form } = 取得選項(設定);
  return 地位[form];
});
```

</details>

**Reply by syimyuzya** [reply]
哇用 class 讓編輯器直接解析其類型好耶w

<details>
<summary>話說這樣的話……</summary>

類型聲明似乎就不是 opt-in 而是預設就有的了？（比如寫了個屬性是 `someValue = "orz"`，會被默認為是 `string`，反而要寫出說它可以是 `any` 才可以 opt-out 的吧？）

而且動態選項的場合下，不同條件下同一個參數有不同類型（某些情況下提供若干 string 作可選值，某些情況下又允許它是任意 string，還有某些場合下它又是個 boolean）又該怎麼辦呢？

以及考慮擴展性：參數順序、分隔符、分組之類的該怎樣做呢？以及以後想給參數加上更多屬性又該怎麼辦呢？

例一：比如我上面說的「參數提供的每個選擇項也要能指定 value、text、description」：

```js
// 沒有 value-text 分離時：
[
  ['版本', [2, '2020：切韻擬音 J（2022 新版）', '2022：切韻擬音 L'],
]
...
if (選項.版本 === '2020：切韻擬音 J（2022 新版）') ... // :(

// 有分離時：
[
  ['版本', 'L', [
    ['J2022', '2020：切韻擬音 J（2022 新版）'],
    ['L', '2022：切韻擬音 L'],
  ]],
]
...
if (選項.版本 === 'J2022') ... // :)
```

例二：再比如以後想給參數再擴展個 `hidden` 屬性（效果是雖然 GUI 上不要顯示，但從 JS API 調用時仍然為其填入預設值）什麼的ww

【目前因為沒這功能，切韻拼音為了在 GUI 隱藏設定（只允許 JS 調用，不允許 GUI 用戶用w），不得不退回過去的「手動填充預設值」的時代😂

</details>

**Reply by graphemecluster** [reply]
全都放到 JSDoc 裡😆

> 讓 GUI 一側變豐富卻令 JS API 一側要顧慮的東西增加，就得不償失了

這就是我想把所有與 API 無關的東西（text、description 等）都放到 comment 裡，推導器是有條件 parse comment 的
（但問題是如果 CLI 也要顯示的話就麻煩一點了）

> 參數順序、分隔符、分組之類的該怎樣做呢？

> 以後想給參數再擴展個 `hidden` 屬性

JSDoc directive

---

那麼真正的問題就是你所説的「類型體操」了

**Reply by graphemecluster** [reply]
```ts
class 選項 {
  $is朗讀音 = this.version === '朗讀音';

  /**
   * 版本
   * 方案版本
   * @value {this['專業模式'] extends true ? '朗讀音' : never} 2019：切韻朗讀音
   * 已過時，僅作爲歷史存檔，不建議使用
   * @value {this['專業模式'] extends true ? 'J' : never} 2020：切韻擬音 J（原版）
   * 已過時，僅作爲歷史存檔，不建議使用
   * @value {'JL'} 2020：切韻擬音 J（2022 新版）
   * @value {'L'} 2022：切韻擬音 L
   */
  version = 'L';

  /**
   * 專業模式
   */
  專業模式 = false;

  // 假設 API 可指定任意字串
  /**
   * @region 基本選項
   * @type {string}
   * @value {this['version'] extends '朗讀音' ? never : 'a'}
   * @value {'ɑ'}
   */
  // 或
  /**
   * @type {string & {} | (this['version'] extends '朗讀音' ? 'ɑ' : 'a' | 'ɑ')}
   */
  後低元音 = !this.$is朗讀音 ? 'a' : 'ɑ';

  /**
   * @type {this['version'] extends '朗讀音' ? 'jɕ' : 'j' | 'ɹ'}
   */
  祭泰夬廢韻尾 = !this.$is朗讀音 ? 'ɹ' : 'jɕ';

  /**
   * 聲調記號
   * @value {['ˊ', 'ˋ']} 上ˊ 去ˋ
   * @value {this['專業模式'] extends true ? ['ʔ', 'h'] : never} 上ʔ 去h
   * @value {['ˀ', 'ʰ']} 上ˀ 去ʰ
   * // 上標的 ʔ Unicode 未收，以 ˀ 代替
   * @value {this['version'] extends '朗讀音' ? never : '五度符號'}
   * @value {this['專業模式'] extends true ? '五度符號（帶拖腔）' : never}
   */
  聲調記號 = ['ˊ', 'ˋ'];

  /**
   * 鈍C介音
   * @value {this['version'] extends 'L' ? null : never} 開∅ 合w
   * @value {this['version'] extends 'JL' ? ['j̈', 'ʋ', 'w'] : never} 開j̈ 唇ʋ 合w
   * @value {this['version'] extends 'JL' ? ['j̈', 'ʋ', 'ɥ̈'] : never} 開j̈ 唇ʋ 合ɥ̈
   * @value {[this['version'], this['專業模式']] extends ['JL' | 'J', true] ? ['j̈', 'β', 'ɥ̈'] : never} 開j̈ 唇β 合ɥ̈
   * @value {this['version'] extends '朗讀音' ? ['j̈', 'ɥ̈', 'ẅ'] : never} 開j̈ 唇ɥ̈ 合ẅ
   */
  鈍C介音 = { L: null, JL: ['j̈', 'ʋ', 'w'], J: ['j̈', 'β', 'ɥ̈'], 朗讀音: ['j̈', 'ɥ̈', 'ẅ'] }[this.version];

  /**
   * @region 高級選項
   * @type {this['專業模式'] extends true ? 'ʈ' | 'tɹ' : never}
   */
  知組 = this.version !== 'J' ? 'ʈ' : 'tɹ';

  /**
   * @type {this['專業模式'] extends true ? this['version'] extends '朗讀音' ? never : 'ˤ' | '˞ ' | '˞' | '͇' | '̳' : never}
   * @value {'ˤ'} 咽化 ◌ˤ
   * @value {'˞ '} r音鉤（帶空隙）◌˞ 
   * @value {'˞'} r音鉤（無空隙）◌˞
   * @value {'͇'} 下等號 ◌͇
   * @value {'̳'} 雙下橫線 ◌̳
   * // 僅在下等號顯示不正常時使用
   */
  二等元音記號 = 'ˤ';

  /**
   * @type {[this['專業模式'], this['version']] extends [true, 'L'] ? boolean : never}
   */
  顯示純音位形式 = false;
}
```

當然這樣的話要做 runtime validation 就比較麻煩，無可厚非

**Comment by graphemecluster** [comment]
折衷版

```ts
const is專業模式 = 選項.專業模式 ?? false;
const is朗讀音 = 選項.version.includes('朗讀音');
const 選項列表 = {
  /**
   * 版本
   * 方案版本
   * @value {'朗讀音'} 2019：切韻朗讀音
   * 已過時，僅作爲歷史存檔，不建議使用
   * @value {'J'} 2020：切韻擬音 J（原版）
   * 已過時，僅作爲歷史存檔，不建議使用
   * @value {'JL'} 2020：切韻擬音 J（2022 新版）
   * @value {'L'} 2022：切韻擬音 L
   */
  version: Option([
    is專業模式 && '朗讀音',
    is專業模式 && 'J',
    'JL',
    'L',
  ], 'L'),

  專業模式: false,

  // 假設 API 可指定任意字串
  /**
   * @region 基本選項
   */
  後低元音: !is朗讀音 ? Option(['a', 'ɑ'], 'a', 'API') : Option(['ɑ'], 'ɑ', 'API'),
  // true: API 和 GUI（「其他，請註明」選項）

  祭泰夬廢韻尾: !is朗讀音 ? Option(['j', 'ɹ'], 'ɹ') : Option(['jɕ'], 'jɕ'),

  /**
   * 聲調記號
   * @value {['ˊ', 'ˋ']} 上ˊ 去ˋ
   * @value {['ʔ', 'h']} 上ʔ 去h
   * @value {['ˀ', 'ʰ']} 上ˀ 去ʰ
   */
  聲調記號: Option([
    ['ˊ', 'ˋ'],
    is專業模式 && ['ʔ', 'h'],
    ['ˀ', 'ʰ'], // 上標的 ʔ Unicode 未收，以 ˀ 代替
    !is朗讀音 && '五度符號',
    is專業模式 && '五度符號（帶拖腔）',
  ], ['ˊ', 'ˋ']);

  /**
   * 鈍C介音
   * @value {[]} 開∅ 合w
   * @value {['j̈', 'ʋ', 'w']} 開j̈ 唇ʋ 合w
   * @value {['j̈', 'ʋ', 'ɥ̈']} 開j̈ 唇ʋ 合ɥ̈
   * @value {['j̈', 'β', 'ɥ̈']} 開j̈ 唇β 合ɥ̈
   * @value {['j̈', 'ɥ̈', 'ẅ']} 開j̈ 唇ɥ̈ 合ẅ
   */
  鈍C介音: {
    L: Option([[]], []),
    JL: Option([['j̈', 'ʋ', 'w'], ['j̈', 'ʋ', 'ɥ̈'], ['j̈', 'β', 'ɥ̈']], ['j̈', 'ʋ', 'w']),
    J: Option([['j̈', 'β', 'ɥ̈']], ['j̈', 'β', 'ɥ̈']),
    朗讀音: Option([['j̈', 'ɥ̈', 'ẅ']], ['j̈', 'ɥ̈', 'ẅ']),
  }[選項.version],

  ...(is專業模式 ? {
    /**
     * @region 高級選項
     */
    知組: Option(['ʈ', 'tɹ'], 選項.version !== 'J' ? 'ʈ' : 'tɹ'),

    /**
     * @value {'ˤ'} 咽化 ◌ˤ
     * @value {'˞ '} r音鉤（帶空隙）◌˞ 
     * @value {'˞'} r音鉤（無空隙）◌˞
     * @value {'͇'} 下等號 ◌͇
     * @value {'̳'} 雙下橫線 ◌̳
     * // 僅在下等號顯示不正常時使用
     */
    二等元音記號: !is朗讀音 ? Option(['ˤ', '˞ ', '˞', '͇', '̳'], 'ˤ') : null,

    顯示純音位形式: 選項.version === 'L' ? false : null,
  } : null),
}
```

**Reply by syimyuzya** [reply]
好耶～

**Reply by syimyuzya** [reply]
此外還有一些功能比如：

- 依當前選項，動態調整參數順序、分組標籤的文本等
- 依當前選項，動態計算參數預設值（和/或可選值）【比如 `['某動態參數', [1, gcd(選項.某參數1, 選項.某參數2)]]`（草🌿），印象中這種操作在 JS 中很 trivial，但拿類型體操寫起來就會很不得了（當然，雖然不是不能寫😂（（
- 給參數指定了類型，要求 GUI 上僅可用該類型；但實則在 JS API 上允許用戶傳入別的指定類型作為高級功能（

在想這些該怎樣做呢ww

**Comment by syimyuzya** [comment]
在想如果真的要給 GUI 編輯器的 `選項` 加自動補全/類型不匹配提示的話，讓（需要它的）用戶直接給方案附加一個形如 `type 選項 =  { 參數1: ...; 參數2: ...; } & Record<string, any>` 的聲明，甚至還更容易些 (?)【固然這樣會導致 JS 和 TS 兩方各寫一遍，有不一致的可能，但考慮到 TS 的類型系統本身就基於「君子協定」……😂
